### PR TITLE
Implemtation of OTA Update and show firmware version of device. Fixes #184

### LINKED ISSUE #184: Add support for OTA updates
Add support for new Z2M OTA feature...
https://github.com/Koenkk/zigbee2mqtt/issues/2921

### PR DESCRIPTION
**Pull request by magpern**
Fixes #184 
Implements OTA Update and will show device firmware verson

If the device is OTA capable and zigbee2mqtt has flagged it as update_available a new option will present

If the device can update this will show:
firmware_available

If the device is up to date, this will show:
firmware_is_uptodate

If the device is not OTA updatable. Nothing will show.
If device firmware verson is available, it will show, else it will say "unknown"

The OTA update option will be visible when the device has reported its status ti zigbee2mqtt.
So, when starting Zigbe2mqttAssistant, no OTA update will show, but after some time it will be available (if device supports it, and zigbee2mqtt supports that device)

### CODE REVIEW

**Comment by carldebilly** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Zigbee2MqttAssistant/Services/BridgeStateService.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/yllibed/Zigbee2MqttAssistant/pullRequest?prid=5086977&bid=16736136)

File: Zigbee2MqttAssistant/Services/BridgeStateService.cs [hunk 1]
```diff
@@ -87,6 +87,7 @@ public ZigbeeDevice UpdateDevice(string friendlyName, string jsonPayload, out bo
 			var json = JObject.Parse(jsonPayload);
 			var linkQuality = json["linkquality"]?.Value<ushort>();
 			var battery = json["battery"]?.Value<decimal>();
+			var updateAvailable = json["update_available"]?.Value<bool?>();
```
Comment (carldebilly) [inline Zigbee2MqttAssistant/Services/BridgeStateService.cs]: Codacy Issue found: [Remove this useless assignment to local variable 'updateAvailable'.](https://app.codacy.com/gh/yllibed/Zigbee2MqttAssistant/pullRequest?prid=5086977)

**Comment by carldebilly** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Zigbee2MqttAssistant/Services/BridgeStateService.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/yllibed/Zigbee2MqttAssistant/pullRequest?prid=5086977&bid=16736136)

**Comment by magpern** [comment]
The variable, updateAvailable is used, so cannot be removed, but I have moved it closer to usage.
Now Visual Studio is no longer complaining about the "useless assignment"
Visual Studio also suggested to invert if-statements to reduce cognitive complexity. So I did that.

File: Zigbee2MqttAssistant/appsettings.json [hunk 2]
```diff
@@ -6,8 +6,10 @@
...
   "AllowedHosts": "*",
   "settings": {
-    //"MqttServer": "mqtt",
-    //"MqttUsername": "",
-    //"MqttPassword": ""
+    "MqttServer": "192.168.1.224",
+    "MqttPort": 1883, 
+    "MqttUsername": "",
+    "MqttPassword": "",
+    "Basetopic":  "zigbee2mqtt2"
```
Comment (carldebilly) [inline Zigbee2MqttAssistant/appsettings.json]: You should remove this from your commit.
Tip: you can locally create a appsettings.Development.json to override those values without modifying the "original" file.

**Comment by carldebilly** [comment]
@all-contributors please add magpern to code

**Comment by carldebilly** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Zigbee2MqttAssistant/Services/BridgeStateService.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/yllibed/Zigbee2MqttAssistant/pullRequest?prid=5086977&bid=16736136)

File: Zigbee2MqttAssistant/Views/Home/Device.cshtml [hunk 3]
```diff
@@ -59,29 +59,62 @@ else
...
+    <div class="card">
+        <div class="card-body">
+            <h5 class="card-title">OTA update</h5>
+            @if (device.IsOtaAvailable == true)
+            {
+                <div class="badge badge-secondary">
+                    Firmware update is available
+                </div>
+                <div class="alert-info">
+                    You can Over-The-Air update you device. Use it wisely, it generates alot of network traffic.
```
Comment (dr1rrb) [inline Zigbee2MqttAssistant/Views/Home/Device.cshtml]: Really nice feature!

I'm doing a useless review, but I noticed this on the screenshot you put in the PR :)

```suggestion
                    You can Over-The-Air update your device. Use it wisely, it generates a lot of network traffic.
```

**Comment by carldebilly** [comment]
@all-contributors please add dr1rrb to code

**Comment by carldebilly** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Zigbee2MqttAssistant/Services/BridgeStateService.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/yllibed/Zigbee2MqttAssistant/pullRequest?prid=5086977&bid=16736136)

**Comment by carldebilly** [comment]
@magpern  I merged it to the `dev` branch to test it on my side.  I'm not sure it's production-ready yet.. but don't hesitate to create new PRs to improve it!

File: Zigbee2MqttAssistant/appsettings.json (hunk 2 above)
Comment (magpern) [reply Zigbee2MqttAssistant/appsettings.json]: My bad.. I really thought I unstaged this before the commit.
I am so new to github. At work we use git and TFS. Fun to learn something new!

**Comment by magpern** [comment]
> @magpern I merged it to the `dev` branch to test it on my side. I'm not sure it's production-ready yet.. but don't hesitate to create new PRs to improve it!

I havn't noticed anything that could break yet. However, it is quite a cunning feature, when you see that an update is available and you press the button, it sends the mqtt request, then redirects to the status page (like the other buttons do).
Then you notice nothing. You have to look at the Z2M log to know what is going on. If you try and do a new update, it will probably cancel the first.

What I would like is for Z2MA to capture the `zigbee2mqtt/bridge/log` and post it to a `Log` page.
The `zigbee2mqtt/bridge/log` shows the update progress (along with other log data, so a filter would be nice). This update progress could also be used to prevent the user from pressing the update again, or notify the user about a failed update. The latter is probably the most common, since all IKEA devices require a power cycle to update, else "update failed"

But to capture the log and post it to a page or an intermediate database, and act on the result, is at this stage a little over my head.

**Comment by carldebilly** [comment]
@magpern I'm a french-speaking canadian, so english is not my language too...

You'll like this: 
https://github.com/Koenkk/zigbee2mqtt/pull/2936#pullrequestreview-357498110  It's merged and already integrated into Z2M `1.11.0`!

For the capture of the log itself (not OTA-related), it's something I already did on a personal branch of Z2M, but it's not finished yet. Still have to create a connected page with the log in Z2MA, probably using SignalR.  Are you interested to investigate that?

We can create an issue to discuss it instead of here :-)

**Comment by magpern** [comment]
@carldebilly SignalIR sounds like a good way to go. I would love to be a part of it, but the web environment is very new to me. I have mostly developed backend and winforms fromtend, but I learn, and the best way is actual use cases :)

**Comment by carldebilly** [comment]
I would love to build something using WebAssembly (I'm one of the architect behind http://platform.uno/ project)... but I think it's not appropriated for such tool :-(  So, right now, I think javascript + SignalR is the best way to go.

**Comment by carldebilly** [comment]
@magpern Let's continue the discussion on this issue: https://github.com/yllibed/Zigbee2MqttAssistant/issues/121

**Pull request merged**